fedbiomed/node/requests/_n2n_controller.py

```diff
@@ -126,7 +134,49 @@ async def _HandlerDefault(self, overlay_msg: dict, inner_msg: InnerMessage) -> N
             f"{ErrorNumbers.FB324}: Failed processing overlay message, unknown inner command "
             f"{inner_msg.__class__.__name__}. Do nothing.")
 
-    async def _HandlerKeyRequest(self, overlay_msg: dict, inner_msg: InnerMessage) -> dict:
+    async def _HandlerChannelRequest(self, overlay_msg: OverlayMessage, inner_msg: InnerMessage) -> dict:
+        """Handler called for ChannelSetupRequest message.
+
+        Args:
+            overlay_msg: Outer message for node to node communication
+            inner_msg: Unpacked inner message from the outer message
+
+        Returns:
+            A `dict` with overlay reply message
+        """
+        # we assume the data is properly formatted
+        inner_resp = ChannelSetupReply(
+            request_id=inner_msg.request_id,
+            node_id=environ['NODE_ID'],
+            dest_node_id=inner_msg.node_id,
+            public_key=await self._overlay_channel.get_local_public_key(inner_msg.node_id))
+
+        overlay, salt = await self._overlay_channel.format_outgoing_overlay(
+            inner_resp,
+            overlay_msg.researcher_id,
+            True
+        )
+        overlay_resp = OverlayMessage(
+            researcher_id=overlay_msg.researcher_id,
+            node_id=environ['NODE_ID'],
+            dest_node_id=inner_msg.node_id,
+            overlay=overlay,
+            setup=True,
+            salt=salt)
+
+        return { 'overlay_resp': overlay_resp }
+
+
+    async def _FinalChannelReply(self, inner_msg: InnerMessage) -> None:
+        """Final handler called for ChannelSetupReply message.
+
+        Args:
+            inner_msg: received inner message
+        """
+        await self._overlay_channel.set_distant_key(inner_msg.get_param('node_id'), inner_msg.get_param('public_key'))
+
+
+    async def _HandlerKeyRequest(self, overlay_msg: OverlayMessage, inner_msg: InnerMessage) -> dict:
```

Comment:
This would be an option, and this is what I did in the very first version (intermediate commit of initial n2n router implementation).
But this means that we have to duplicate the `format_incoming_overlay` in each callback, while it could be factored in the calling `_NodeToNodeAsyncRouter._overlay_message_process`.
Moreover, you suggested that when we refactor n2n controller, we move the building of the reply `OverlayMessage` out of the callback (so maybe in `_overlay_message_process`). 
So in the end we may rather go, when refactoring, towards a call with only the `InnerMessage`:

async def _HandlerKeyRequest(self, inner_msg: InnerMessage) -> dict:

Anyway, this was not changed in this PR, so I think changing this would rather belong to the refactoring of the N2n router part than to this PR ...